internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -334,9 +334,20 @@ func (r *BMACReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (c
 
 // handleBMHFinalizer adds a finalizer to the BMH if it isn't being deleted and the finalizer isn't already present
 // If the BMH is being deleted, the associated node is drained and the agent (if any) is annotated and deleted before the BMH is allowed to be removed
-// Finalizer handling is only run if the user has annotated the BMH and converged flow is enabled
+// Finalizer handling is only run if the user has annotated the BMH and converged flow is enabled,
+// or, when the BMH is annotated with 'paused' annotation.
 func (r *BMACReconciler) handleBMHFinalizer(ctx context.Context, log logrus.FieldLogger, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent) reconcileResult {
-	if _, has_annotation := bmh.GetAnnotations()[BMH_DELETE_ANNOTATION]; !has_annotation || !r.ConvergedFlowEnabled {
+	var removeAgentOnDelete bool
+	if _, has_annotation := bmh.GetAnnotations()[BMH_DELETE_ANNOTATION]; has_annotation && r.ConvergedFlowEnabled {
+		removeAgentOnDelete = true
+	}
+
+	var bmhPaused bool
+	if _, has_annotation := bmh.GetAnnotations()[BMH_PAUSED_ANNOTATION]; has_annotation {
+		bmhPaused = true
+	}
+
+	if !removeAgentOnDelete && !bmhPaused {
```

Comment:
How come we don't set a finalizer if it's missing both the remove agent annotation and the paused annotation? :o

Reply:
The two situations where we have work to do when the BMH is deleted are:
1. The user set the remove agent on delete annotation, so we need to deprovision then remove the agent.
2. The BMH is paused, so we need to unpause the BMH to allow BMO to finalize it.

If neither is set there's no work for us to do so I don't think we need a finalizer.

That all said, might indeed be safer to just set it unconditionally then make the cleanup work conditional on what the resource looks like when it's deleted.

I mention that because we're not removing the finalizer if the BMH used to be paused but it now (when it is deleted) not, right? I could see that becoming a bug. And this is probably not the only problem. Surely there are more corner cases like this that I haven't thought of.

Reply:
Ah I see that we don't need to do work if there's no pause/delete annotation, thanks for the explanation! I agree that we should set it unconditionally too since there could be other cases like what you've mentioned.

Reply:
Good points, will set in unconditionally then.